### ISSUE TITLE
Trainable sentence segmenter

### ISSUE THREAD

**Issue by grayfall** [skoblov-lab/SciLK#6]

**Comment by grayfall** [comment]
@edikedik get a set of all punctuation marks in Genia. Get a subset of sentence-terminating marks.

**Comment by edikedik** [comment]
```
import re
from typing import Tuple, Set, Text
from functools import reduce

path = 'local/GENIAcorpus3.02.xml'
parsed = genia.parse(path)

PATTERN = re.compile("""(?x)                   # set flag to allow verbose regexps
                    (?:[A-Z]\.)+           # abbreviations, e.g. U.S.A.
                    |\d+(?:\.\d+)?%?       # numbers, incl. currency and percentages
                    |\w+(?:[-']\w+)*       # words w/ optional internal hyphens/apostrophe
                    |(?:[/\-@&])         # special characters with meanings""")

# PATTERN is intended to capture all the chars between valid words (which include hyphenated words,
# abbreviations, etc.)

def get_punctuation(abstract) -> Tuple[Set[Text], Set[Text]]:
    text = abstract[0][2]
    sent_borders = abstract[2][2]
    # get characters between sentences
    bw_sent_int = [
        (prev_int.stop - 1, next_int.start - 1) if next_int.start is not None else (prev_int.stop - 1, prev_int.stop)
        for prev_int, next_int in zip(sent_borders, sent_borders[1:] + [Interval(None, None, None)])]
    bw_sent_char = set(text[start: stop] for start, stop in bw_sent_int)
    # get all the characters between PATTERN matches 
    in_sent_char = set(PATTERN.split(text))
    return bw_sent_char, in_sent_char


punctuation = list(map(get_punctuation, parsed))

bw_sent = set("".join(reduce(lambda x, y: x | y, (x[0] for x in punctuation))))
in_sent = set("".join(reduce(lambda x, y: x | y, (x[1] for x in punctuation))))
```

> get a set of all punctuation marks in Genia

`{':', '?', ' ', ')', '(', '*', '=', '+', '.', '>', '[', ',', ';', '<', ']', "'", '"'}`
> Get a subset of sentence-terminating marks

` {' ', '"', '.', '?'}`

**Comment by grayfall** [comment]
@edikedik can you show some sentences terminated by `' '` and `'"'`. Attach their abstract IDs and source type (title/body): these data might prove useful for further investigations.

**Comment by edikedik** [comment]
@grayfall 
Actually, after I fixed minor bug in my function, the whole set of sentence terminating marks turned out to be bigger.

`{'n', 'l', 'B', ')', 't', 'o', 's', '%', 'r', 'C', '?', 'm', '.', 'y', 'a', '2', 'd', '"', 'h', '1'}`

In virtually all of the cases I looked into unconventional endings are situated in the sentence, terminating abstract body, i.e. there is no `.` at the end of a sentence.

Attaching .tsv with PMIDs and unconventional endings (considering only `?` and `.` to be conventional).
There are 65 of them in total.
[unconventional_body_endings.zip](https://github.com/skoblov-lab/SciLK/files/1431016/unconventional_body_endings.zip)
